ateam_ai/src/trajectory_generation/ilqr_computer.hpp

```diff
@@ -0,0 +1,346 @@
+      // eq 4c
+      Q_xx = l_xx + f_x_T * V_xx * f_x;
+      // eq 4d
+      Q_ux = l_ux + f_u_T * V_xx * f_x;
+      // eq 4e
+      Q_uu = l_uu + f_u_T * V_xx * f_u;
+
+      // eq 5b
+      // Solve inverse with regulization to keep it from exploding
+      Eigen::Matrix<double, U, U> Q_uu_inv = inverse(Q_uu);
+
+      k.at(t) = -1 * Q_uu_inv * Q_u;
+      K.at(t) = -1 * Q_uu_inv * Q_ux;
+
+      // eq 6a
+      // Eigen doesn't like converting a 1x1 matrix to double :(
+      delta_v.at(t) = -0.5 * (k.at(t).transpose() * Q_uu * k.at(t)).value();
+      // eq 6b
+      V_x = Q_x - K.at(t).transpose() * Q_uu * k.at(t);
+      // eq 6c
+      V_xx = Q_xx - K.at(t).transpose() * Q_uu * K.at(t);
+    }
+
+    return std::make_tuple(k, K, delta_v);
+  }
+
+  std::pair<Trajectory, Actions> ammend_trajectory_actions(
+    const Trajectory & reference_traj,
+    const Actions & reference_action,
+    const Feedforwards & k,
+    const Feedbacks & K,
+    const double alpha)
+  {
+    // Eq 7a/b/c
+    Trajectory candidate_trajectory;
+    Actions candidate_actions;
+
+    candidate_trajectory.front() = trajectory.front();
+    for (Time t = 0; t < T - 1; t++) {
+      const State & x_hat_i = candidate_trajectory.at(t);
+      const State & x_i = trajectory.at(t);
+      const Input & u_i = actions.at(t);
+      const Input & u_hat_i = u_i + alpha * k.at(t) + K.at(t) * (x_hat_i - x_i);
+      candidate_actions.at(t) = u_hat_i;
+      candidate_trajectory.at(t + 1) = problem.dynamics(x_hat_i, u_hat_i, t);
+    }
+
+    return std::make_pair(candidate_trajectory, candidate_actions);
+  }
+
+  void decrease_regulation()
+  {
+    alpha *= params.alpha_change;
+  }
+
+  void increase_regulation()
+  {
+    alpha /= params.alpha_change;
+  }
+
+  inline StateJacobian_x state_jacobian_x(const StateJacobian & j)
+  {
+    return j.block(0, 0, X, X);
+  }
+  inline StateJacobian_u state_jacobian_u(const StateJacobian & j)
+  {
+    return j.block(0, X, X, U);
+  }
+
+  inline CostGradiant_x cost_gradiant_x(const CostGradiant & g)
+  {
+    return g.block(0, 0, X, 1);
+  }
+  inline CostGradiant_u cost_gradiant_u(const CostGradiant & g)
+  {
+    return g.block(X, 0, U, 1);
+  }
+
+  inline CostHessian_xx cost_hessian_xx(const CostHessian & h)
+  {
+    return h.block(0, 0, X, X);
+  }
+  inline CostHessian_ux cost_hessian_ux(const CostHessian & h)
+  {
+    return h.block(X, 0, U, X);
+  }
+  inline CostHessian_uu cost_hessian_uu(const CostHessian & h)
+  {
+    return h.block(X, X, U, U);
+  }
+
+  inline Eigen::Matrix<double, U, U> inverse(const Eigen::Matrix<double, U, U> & m)
+  {
+    // Solve inverse with regulization to keep it from exploding
+    Eigen::SelfAdjointEigenSolver<Eigen::Matrix<double, U, U>> eigensolver(m);
+    if (eigensolver.info() != Eigen::Success) {
+      std::cout << "Failed to solve eigen vectors" << std::endl;
+      // TODO(jneiger): Replace with tikhonov regularization + inverse which can't fail
+      return m;
```

Comment:
I already have the a different regularization + inverse method implemented that I didn't want to clutter this PR with. Will swap to that in a later PR so just doing the bare minimum to compile rn